File: src/game/state/Combat.cpp
//
// Created by Per-Arne on 27.02.2017.
//

#include "Combat.h"
#include "../unit/Unit.h"
#include "../player/Player.h"
#include "../util/Pathfinder.h"


void Combat::update(Unit & unit)const{

    unit.combatTimer += 1;
    if(unit.combatTimer >= unit.combatInterval) {
        Unit &combatTarget = unit.getCombatTarget();


		// If combat target are despawn in some way (Building, Dead, Despawned) stop attacking...
		if (!combatTarget.tile) {
			unit.combatTargetID = -1;
			unit.combatTimer = 1000;
			unit.transitionState();
			return;
		}

        if(unit.distance(*combatTarget.tile) > unit.damageRange) {
            // Too far away, Walk

			unit.walkingGoalID = combatTarget.tile->id;
			unit.transitionState(unit.stateManager->walkingState);
            unit.enqueueState(unit.stateManager->combatState);

        } else {
            // Can attack
            int myDamage = unit.getDamage(combatTarget);
            combatTarget.afflictDamage(myDamage);
            unit.player_->statUnitDamageDone += myDamage;
            combatTarget.player_->statUnitDamageTaken += myDamage;
            unit.combatTimer = 0;

            if(combatTarget.isDead()){
                unit.combatTargetID = Constants::None;
                unit.combatTimer = 1000;
                unit.transitionState();
                return;
            }

            if(combatTarget.state->id == Constants::State::Idle) {
                combatTarget.attack(*unit.tile);
            }

        }


    }

}

void Combat::end(Unit & unit)const{

}

void Combat::init(Unit & unit)const{
    Position dir = unit.distanceVector(*unit.getCombatTarget().tile);
    unit.setDirection(dir);



}

File: python/game/state/Combat.py
from game import Config
from game import Constants


class Combat:
    name = "Combat"
    id = Constants.State.Combat

    def __init__(self):
        pass

    def update(self, unit):
        self.data['attack_timer'] += tick

        if self.data['attack_timer'] >= self.attack_interval:
            attack_target = self.game.units[self.data['attack_target']]

            distance = self.unit.distance(
                attack_target.d['x'],
                attack_target.d['y'],
                attack_target.dimension
            )
            if distance > self.unit.range:
                # Too far away, Walk

                # Find adjacent tile to the attack target
                tiles = self.game.AdjacentMap.adjacent_walkable(
                    self.game,
                    attack_target.d['x'] + attack_target.dimension,
                    attack_target.d['y'] + attack_target.dimension,
                    attack_target.dimension + 1
                )
                # No tiles are available around the target
                if not tiles:
                    return

                # Find the tile with shortest distance
                tile = self.unit.shortest_distance(tiles)

                # Next state is Combat (After movement)
                self.add_next(Combat.type, {
                    'attack_target': self.data['attack_target']
                }, 1)

                # Move to tile (This triggers transition)
                self.unit.move(*tile)
            else:
                # Can attack (Distance < Range)
                my_damage = self.unit.get_damage(self.unit)
                attack_target.afflict_damage(my_damage)

                # If attack target is dead, transition to next state
                if attack_target.is_dead():
                    self.data['attack_target'] = None
                    self.player.statistics['kill_count'] += 1
                    self.transition()
                    return

                # Make unit retaliate if its Idle (Changes to attack state)
                if attack_target.state.type == const.State.ID_Idle:
                    attack_target.attack(self.unit.d['x'], self.unit.d['y'])

            self.data['attack_timer'] = 0

    def end(self, unit):
        pass

    def init(self, unit):
        pass